### PR TITLE
experimental: repair index endpoint

### PR DESCRIPTION
**Pull request by asdine**
### What's being changed:

This adds a new debug/experimental endpoint to repair indexes.

Usage:

```bash
curl -I -v -X POST http://localhost:6060/debug/index/repair/vector?shard=foo&collection=bar&vector=baz
```

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/shard_read.go [hunk 1]
```diff
@@ -271,6 +271,9 @@ func (s *Shard) getIndexQueue(targetVector string) (*IndexQueue, error) {
 		}
 		return queue, nil
 	}
+	if targetVector != "" {
+		return nil, fmt.Errorf("index queue: target vector not found: %q", targetVector)
```
Comment (antas-marcin) [inline adapters/repos/db/shard_read.go:275]: this method was returning a queue for `targetVector == ""` if there were no targetVectors (named vectors configured, see line 264), now this method changed to handle named vectors, there's no fallback to returning a queue when there's no target vector defined, is that OK?

File: adapters/repos/db/index_queue.go [hunk 2]
```diff
@@ -473,6 +476,8 @@ func (q *IndexQueue) pushToWorkers(max int, wait bool) int64 {
 			}
 		}
 
+		q.queue.ResetDeleted(deleted...)
```
Comment (antas-marcin) [inline adapters/repos/db/index_queue.go:479]: I can see this change in the `pushToWorkers` method, this won't break the existing logic, yes?

File: adapters/repos/db/shard_read.go (hunk 1 above)
Comment (asdine) [reply adapters/repos/db/shard_read.go:275]: Yes, I figured it's safer to return an error when we are calling this method expecting a named vector when there are none.
Before this change, we could potentially end up pushing to the wrong queue.

File: adapters/repos/db/index_queue.go (hunk 2 above)
Comment (asdine) [reply adapters/repos/db/index_queue.go:479]: No it should not, this ensures we always clean up the blacklist before dequeuing

File: adapters/repos/db/shard_read.go (hunk 1 above)
Comment (antas-marcin) [reply adapters/repos/db/shard_read.go:275]: ok, then if this method is meant to be used only with target vectors, bc now with this check in place we will never reach line 227 (`return s.queue, nil`) yes?

Comment (asdine) [reply adapters/repos/db/shard_read.go:275]: > we will never reach line 227 (return s.queue, nil) yes?

Actually not exactly, this method is there to simplify getting the right queue depending on the configuration of the shard: here's an example:

```go
// Example 1: no named vectors
s.getIndexQueue("") // valid call: this reaches l.277

// Example 2: no named vectors
s.getIndexQueue("foo") // invalid call: this reaches l.275

// Example 3: named vectors
s.getIndexQueue("") // invalid call: this reaches l.266

// Example 4: named vectors
s.getIndexQueue("foo") // valid call: this reaches l.272
```

Does that make sense?

Comment (antas-marcin) [reply adapters/repos/db/shard_read.go:275]: yeah, now I see it

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**